core/src/main/scala/cats/data/EitherT.scala

```diff
@@ -90,11 +90,26 @@ final case class EitherT[F[_], A, B](value: F[Either[A, B]]) {
   def flatMapF[AA >: A, D](f: B => F[Either[AA, D]])(implicit F: Monad[F]): EitherT[F, AA, D] =
     flatMap(f andThen EitherT.apply)
 
+  def subflatMap[AA >: A, D](f: B => Either[AA, D])(implicit F: Functor[F]): EitherT[F, AA, D] =
+    transform(_.flatMap(f))
+
+  def semiflatMap[D](f: B => F[D])(implicit F: Monad[F]): EitherT[F, A, D] =
+    flatMap(b => EitherT.right(f(b)))
+
   def transform[C, D](f: Either[A, B] => Either[C, D])(implicit F: Functor[F]): EitherT[F, C, D] =
     EitherT(F.map(value)(f))
 
-  def subflatMap[AA >: A, D](f: B => Either[AA, D])(implicit F: Functor[F]): EitherT[F, AA, D] =
-    transform(_.flatMap(f))
+  def flatTap[AA >: A, D](f: B => EitherT[F, AA, D])(implicit F: Monad[F]): EitherT[F, AA, B] =
+    flatMap(b => f(b).map(_ => b))
+
+  def flatTapF[AA >: A, D](f: B => F[Either[AA, D]])(implicit F: Monad[F]): EitherT[F, AA, B] =
+    flatTap(f andThen EitherT.apply)
+
+  def subflatTap[AA >: A, D](f: B => Either[AA, D])(implicit F: Functor[F]): EitherT[F, AA, B] =
+    subflatMap(b => f(b).map(_ => b))
```

Comment:
`.as(b)` can be optimized and by law should be the same as `.map(_ => b)`.

For instance, as for `EitherMonad` can be a cast of the type if we have a `Left` or allocate `Right(b)`, avoiding allocating the lambda and calling a function (which is hard to inline for the JIT due to Function1 being megamorphic).

Reply:
@johnynek There are many `.map(_ => a)` here and there in my change. To do the optimization you suggest I need to implement `as` method for each class changed, override `as` in their `Functor` instances, express `flatTap`/`subflatTap` via `as` instead of `map`. For `Either`/`Ior` `leftAs` should be added for the same purpose. Right?

Reply:
I would say for this PR express `flatTap/subflatTap` via `as` instead of `map` would suffice. We can create a separate PR or maybe just an issue for optimizing `as`.

Reply:
Ok, I will rework this PR.